:warning: Bump to Go 1.13

joelanford opened the pull request:
This PR builds on #598 by completing more of the tasks listed in #590:

- [x] Prune any uses of xerrors (89320998b0928b69df3b5fa255a3eb428333213c)
- [x] Use %w instead of %v in Errorf (fc42a0a06fc146f0a1d314ca254160409617295d)
- [x] Prune uses of GO111MODULE=on from our scripts (e0a1fbb718d55f200c27d2a7dfde7f91901cd8ec)

(note: we already did a soft-bump to 1.13 the previous release, this just starts using the features)

alvaroaleman commented:
/ok-to-test

In order for the tests to use go 1.13, the image needs to be bumped here: https://github.com/kubernetes/test-infra/blob/e0a1e8e0d0f947087e7b8295f49b21d772f7a070/config/jobs/kubernetes-sigs/controller-runtime/controller-runtime-presubmits.yaml#L13

DirectXMan12 commented:
(removed my at-mention from the PR description -- that causes all sorts of issues when it merges)

rajathagasthya commented on hack/check-everything.sh:
```diff
@@ -70,7 +70,7 @@ function fetch_go_tools {
   header_text "Checking for gometalinter.v2"
```
nit: while you are changing this file, `s/gometalinter.v2/golangci-lint/`

rajathagasthya commented on hack/check-everything.sh:
```diff
@@ -70,7 +70,7 @@ function fetch_go_tools {
   if ! is_installed golangci-lint; then
     header_text "Installing golangci-lint"
-    GO111MODULE=on go get github.com/golangci/golangci-lint/cmd/golangci-lint@v1.15.0
+    go get github.com/golangci/golangci-lint/cmd/golangci-lint@v1.15.0
```
I know this isn't changing current behavior, but maybe this tool should be installed with `GO111MODULE=off`? The current version will add this as a dependency to `go.mod` cluttering the file. We don't really want to carry a lint tool as a dependency and someone might accidentally commit it.

Related https://github.com/golang/go/issues/30515

joelanford replied:
I agree that it isn't great to clutter and inadvertantly change `go.mod`. This approach doesn't work though:

```
$ ./hack/check-everything.sh
using tools
Checking for golangci-lint
Installing golangci-lint
go: cannot use path@version syntax in GOPATH mode
```

I updated this to use the `curl`/`goreleaser` installation method that's used elsewhere. Was there some reason we weren't using that here?

rajathagasthya replied:
> The GO111MODULE=off approach doesn't work though

Ah, yes, because of the version. Got it. 

> Was there some reason we weren't using that here?

If I have to guess, it was done before go modules were introduced and it never changed.

joelanford commented:
No huge rush here, but this is still waiting on https://github.com/kubernetes/test-infra/pull/14373.

I _think_ that should pass CI and be able to be merged without breaking `master` on controller-runtime, but I'm not positive.

@DirectXMan12 can you add `ok-to-test` to https://github.com/kubernetes/test-infra/pull/14373 when you get a chance?

alvaroaleman commented:
@joelanford Does it make sense to have two presubmits, one with go 1.12 for the release-0.2 branch and one with go 1.13 for the master branch or can we safely assume that the presubmit with go 1.13 will work on both?

joelanford commented:
@alvaroaleman I see no reason it would not work on both. But the question in my mind is whether this PR will necessitate moving to `v0.3.0` (I think not, but I'm not sure).

If we jump to `v0.3.0` after this merges, it would probably make sense to continue using go1.12 for the `release-0.2` branch.

If we just continue to `v0.2.3` after this merges, then the next commit on `release-0.2` will need go1.13, so I think there would be no reason to have a separate presubmit, unless there's a chance that we backport a bug fix to `release-0.2` without releasing everything else on `master`.

DirectXMan12 commented:
We *probably* want to make sure we don't break go1.12 on release-0.2 -- IMO, we'll want to go to v0.3.0 for go1.13, but I'm not certain exactly how we want to deal with language version bumps.

DirectXMan12 commented:
What do other people think, especially end consumers of CR?  Is go-1.13 in all the major distros at this point yet?

joelanford commented:
Just updated kubernetes/test-infra#14373 to have separate builds for each branch and to bump `master` to use go 1.13.

I also updated this PR to move from `xerrors` (which we introduced in the DynamicRESTMapper PR) to `errors`.

Anything else we should handle here?

DirectXMan12 commented:
/lgtm
/approve
